data/src/main/java/org/apache/iceberg/data/GenericReader.java

```diff
@@ -92,67 +70,16 @@
   }
 
   public CloseableIterable<Record> open(FileScanTask task) {
-    List<DeleteFile> posDeletes = Lists.newArrayList();
-    List<DeleteFile> eqDeletes = Lists.newArrayList();
-    for (DeleteFile delete : task.deletes()) {
-      switch (delete.content()) {
-        case POSITION_DELETES:
-          posDeletes.add(delete);
-          break;
-        case EQUALITY_DELETES:
-          eqDeletes.add(delete);
-          break;
-        default:
-          throw new UnsupportedOperationException("Unknown delete file content: " + delete.content());
-      }
-    }
+    DeleteFilter<Record> deletes = new GenericDeleteFilter(io, task, tableSchema, projection);
+    Schema readSchema = deletes.requiredSchema();
 
-    Schema fileProjection = fileProjection(posDeletes, eqDeletes);
-
-    CloseableIterable<Record> records = openFile(task, fileProjection);
-    records = applyPosDeletes(records, fileProjection, task.file().path(), posDeletes, task.file());
-    records = applyEqDeletes(records, fileProjection, eqDeletes, task.file());
-    records = applyResidual(records, fileProjection, task.residual());
+    CloseableIterable<Record> records = openFile(task, readSchema);
```

Comment:
Q:  since we've added few columns which may not be included in `projection` schema  in the `readScheme`, then  Will the record from iterator may have more columns than user expected ?  

For example,  table test=(a,b,c),  the user query the data by `select b from test where b >10`,  while we use `a` as the equality field id set for equality delete files,  this `CloseableIterable<Record>` will return records with column `(a,b)` , while users would actually expected return records with column `(b)` ?

Reply:
Yes, the records will have more columns than requested. That's why we project the added columns at the end of the record. @shardulm94 pointed this out on the original PR for generics.

For Spark, this is okay because Spark ignores the extra columns if the schema associated with a row doesn't have them. You can see the check here: https://github.com/apache/iceberg/pull/1444/files#diff-7600f4d25cfdef7f5da70e12126b55c7R137-R138

For generics, we just return the larger row to avoid needing to make a copy right now. I don't think it is worth the cost of a copy to remove the columns, so we would need to add the ability to truncate the columns of a `GenericRecord`. I think that adding that feature to `GenericRecord` should be done in a separate PR, if we decide that it _should_ be done.